Marlin/src/module/ft_motion.h

```diff
@@ -181,14 +176,132 @@ class FTMotion {
+                                                                   // scalar in the right hand because types.h is missing scalar on left cases
+          // using halves since we'll double the dividend and iteration_multiplier inside the while.
+          xyze_ulong_t half_space = space >> 1;                         // ensures (dividend_scaled<<1) <= space
+          uint32_t half_pending = (bresenham_iterations_pending - 1) >> 1;   // ensures (iter<<1) < pending
+
+          // —— Grow phase: keep doubling until right before a step would occur (or pending iterations exhausted)
+          while (dividend_scaled <= half_space && jump <= half_pending) {
+            jump <<= 1;               // stays < pending
+            dividend_scaled <<= 1;    // safe: fits after doubling
+          }
+
+          // —— Shrink phase: keep adding and halving until there's 1 iteration left for actual stepping
+          while (bresenham_iterations_pending > 1 && jump > 0) {
+            if (jump < bresenham_iterations_pending && dividend_scaled <= space) {
+              space -= dividend_scaled;
+              bresenham_iterations_pending -= jump;
+              delta_error_q32 += dividend_scaled;
+              iterations += jump;
+            }
+            dividend_scaled >>= 1;
+            jump >>= 1;
+          }
+        }
+
+        step_bits = 0;
+
+        // if there are still iteration pending (should be exactly 1), run through it
+        if (bresenham_iterations_pending) {
+          iterations++;
+          bresenham_iterations_pending--;
+          #define RUN_AXIS(A) do{                                                      \
+              delta_error_q32.A += stepper_plan.advance_dividend_q0_32.A;              \
+              step_bits.A = delta_error_q32.A < stepper_plan.advance_dividend_q0_32.A; \
+            }while(0);
+          LOGICAL_AXIS_MAP(RUN_AXIS);
+          #undef RUN_AXIS
+        }
+
+        // TODO: find out why not subtracting 1 results in 1 bad pulse train per trajectory point.
+        return (iterations - 1) * INTERVAL_PER_ITERATION;
+      }
+
+      /**
+       * If bresenham_iterations_pending, advance to next actual step.
+       * Else, consume stepper data point
+       * Then return interval until that next step
+       */
+      uint32_t plan() {
+        if (bresenham_iterations_pending > 0) return advance_until_step();
+
+        if (stepper_plan_is_empty()) {
+          bresenham_iterations_pending = 0;
+          step_bits = 0;
+          return HAL_TIMER_TYPE_MAX;
+        }
+
+        AxisBits old_dir_bits = stepper_plan.dir_bits;
+        stepper_plan = dequeue_stepper_plan();
+        const AxisBits dir_flip_mask = old_dir_bits ^ stepper_plan.dir_bits;  // axes that must toggle now
+        if (dir_flip_mask) {
+          #define _HANDLE_DIR_CHANGES(A) if (dir_flip_mask.A) delta_error_q32.A *= -1;
```

Comment:
Replacing `delta_error_q32.A = - delta_error_q32.A;` with `delta_error_q32.A *= -1;` may make the compiler use mult instead of negating (cheaper), and this is all unsigned math so *=-1 is less clear.

It's better to revert this change

Reply:
Ok, I read that as long that things are compiled with optimisation flags, it should result in the same machine code.
It's less explicit imo but it should be equivalent

Reply:
It follows on the general principle to use semantics and code features to avoid repetition where possible with the understanding that the GCC compiler does allow for deep unoptimization when needed. We trust that nothing will ship without the compiler applying the most efficient form of negation available on the target platform.